llap-common/src/java/org/apache/hadoop/hive/llap/metrics/ReadWriteLockMetrics.java

```diff
@@ -0,0 +1,428 @@
+import org.apache.hadoop.metrics2.lib.MutableCounterLong;
+
+/**
+ * Wrapper around a read/write lock to collect the lock wait times.
+ * Instances of this wrapper class can be used to collect/accumulate the wait
+ * times around R/W locks. This is helpful if the source of a performance issue
+ * might be related to lock contention and you need to identify the actual
+ * locks. Instances of this class can be wrapped around any <code>ReadWriteLock
+ * </code> implementation.
+ */
+public class ReadWriteLockMetrics implements ReadWriteLock {
+  private LockWrapper readLock;         ///< wrapper around original read lock
+  private LockWrapper writeLock;        ///< wrapper around original write lock
+
+  /**
+   * Helper class to compare two <code>LockMetricSource</code> instances.
+   * This <code>Comparator</code> class can be used to sort a list of <code>
+   * LockMetricSource</code> instances in descending order by their total lock 
+   * wait time.
+   */
+  public static class MetricsComparator implements Comparator<MetricsSource> {
+    @Override
+    public int compare(MetricsSource o1, MetricsSource o2) {
+      if (o1 != null && o2 != null
+          && o1 instanceof LockMetricSource && o2 instanceof LockMetricSource) {
+        LockMetricSource lms1 = (LockMetricSource)o1;
+        LockMetricSource lms2 = (LockMetricSource)o2;
+
+        long totalMs1 = (lms1.readLockWaitTimeTotal.value() / 1000000L)
+                        + (lms1.writeLockWaitTimeTotal.value() / 1000000L);
+        long totalMs2 = (lms2.readLockWaitTimeTotal.value() / 1000000L) 
+                        + (lms2.writeLockWaitTimeTotal.value() / 1000000L);
+
+        // sort descending by total lock time
+        if (totalMs1 < totalMs2) {
+          return 1;
+        }
+
+        if (totalMs1 > totalMs2) {
+          return -1;
+        }
+
+        // sort by label (ascending) if lock time is the same
+        return lms1.lockLabel.compareTo(lms2.lockLabel);
+      }
+
+      return 0;
+    }
+  }
+
+  /**
+   * Wraps a <code>ReadWriteLock</code> into a monitored lock if required by
+   * configuration. This helper is checking the <code>
+   * hive.llap.lockmetrics.collect</code> configuration option and wraps the
+   * passed in <code>ReadWriteLock</code> into a monitoring container if the
+   * option is set to <code>true</code>. Otherwise, the original (passed in)
+   * lock instance is returned unmodified.
+   *
+   * @param conf Configuration instance to check for LLAP conf options
+   * @param lock The <code>ReadWriteLock</code> to wrap for monitoring
+   * @param metrics The target container for locking metrics
+   * @see #createLockMetricsSource
+   */
+  public static ReadWriteLock wrap(Configuration conf, ReadWriteLock lock, 
+                                   MetricsSource metrics) {
+    Preconditions.checkNotNull(lock, "Caller has to provide valid input lock");
+    boolean needsWrap = false;
+
+    if (null != conf) {
```

Comment:
those 2 ifs can be one if (less missed branches). ` conf != null && HiveConf.getBoolVar(conf, HiveConf.ConfVars.LLAP_COLLECT_LOCK_METRICS) ? new ReadWriteLockMetrics(lock, metrics) : lock`